File: correlation_package/correlation_package_lib/corr_cuda.cpp
#include "corr_cuda.h"
#include "corr_cuda_kernel.h"

corr_cuda::corr_cuda()
{}

void corr_cuda::forward(at::Tensor input1,
                      at::Tensor input2,
                      at::Tensor rbot1,
                      at::Tensor rbot2,
                      at::Tensor output,
                      int pad_size,
                      int kernel_size,
                      int max_displacement,
                      int stride1,
                      int stride2,
                      int corr_type_multiply,
                      int batchSize,
                      int nInputPlane,
                      int nInputRows,
                      int nInputCols
                      )
{
    int inputWidthHeight = nInputRows * nInputCols;

    int kernel_radius_ = (kernel_size - 1) / 2;
    int border_size_ = max_displacement + kernel_radius_; // size of unreachable border region (on each side)

    int paddedbottomheight = nInputRows + 2 * pad_size;
    int paddedbottomwidth = nInputCols + 2 * pad_size;

    int nOutputCols = ceil((float)(paddedbottomwidth - border_size_ * 2) / (float)stride1);
    int nOutputRows = ceil((float)(paddedbottomheight - border_size_ * 2) / (float)stride1);

    // Given a center position in image 1, how many displaced positions in -x / +x
    // direction do we consider in image2 (neighborhood_grid_width)
    int neighborhood_grid_radius_ = max_displacement / stride2;
    int neighborhood_grid_width_ = neighborhood_grid_radius_ * 2 + 1;

    // Number of output channels amounts to displacement combinations in X and Y direction
    int nOutputPlane = neighborhood_grid_width_ * neighborhood_grid_width_;

    // Inputs
    float * input1_data = input1.data<float>();
    float * input2_data = input2.data<float>();

    // Outputs
    float * output_data = output.data<float>();

    float * rbot1_data = rbot1.data<float>();
    float * rbot2_data = rbot2.data<float>();

    int pwidthheight = paddedbottomwidth * paddedbottomheight;

    blob_rearrange_ongpu(input1_data,rbot1_data,batchSize,nInputPlane,nInputCols,nInputRows,inputWidthHeight,pad_size,pwidthheight);

    blob_rearrange_ongpu(input2_data,rbot2_data,batchSize,nInputPlane,nInputCols,nInputRows,inputWidthHeight,pad_size,pwidthheight);

    CorrelateData_ongpu(rbot1_data,rbot2_data,output_data,batchSize,nOutputCols,nOutputRows,nOutputPlane,max_displacement,neighborhood_grid_radius_,neighborhood_grid_width_,kernel_radius_,kernel_size,stride1,stride2,paddedbottomwidth,paddedbottomheight,nInputPlane,corr_type_multiply);
}

void corr_cuda::backward(at::Tensor input1,
                        at::Tensor input2,
                        at::Tensor rbot1,
                        at::Tensor rbot2,
                        at::Tensor gradOutput,
                        at::Tensor gradInput1,
                        at::Tensor gradInput2,
                        int pad_size,
                        int kernel_size,
                        int max_displacement,
                        int stride1,
                        int stride2,
                        int corr_type_multiply,
                        int batchSize,
                        int nInputPlane,
                        int nInputRows,
                        int nInputCols)
{

    float * input1_data = input1.data<float>();
    float * input2_data = input2.data<float>();

    float * gradOutput_data = gradOutput.data<float>();
    float * gradInput1_data = gradInput1.data<float>();
    float * gradInput2_data = gradInput2.data<float>();

    int inputWidthHeight = nInputRows * nInputCols;

    int kernel_radius_ = (kernel_size - 1) / 2;
    int border_size_ = max_displacement + kernel_radius_; // size of unreachable border region (on each side)

    int paddedbottomheight = nInputRows + 2 * pad_size;
    int paddedbottomwidth = nInputCols + 2 * pad_size;

    int nOutputCols = ceil((float)(paddedbottomwidth - border_size_ * 2) / (float)stride1);
    int nOutputRows = ceil((float)(paddedbottomheight - border_size_ * 2) / (float)stride1);

    // Given a center position in image 1, how many displaced positions in -x / +x
    // direction do we consider in image2 (neighborhood_grid_width)
    int neighborhood_grid_radius_ = max_displacement / stride2;
    int neighborhood_grid_width_ = neighborhood_grid_radius_ * 2 + 1;

    // Number of output channels amounts to displacement combinations in X and Y direction
    int nOutputPlane = neighborhood_grid_width_ * neighborhood_grid_width_;

    float * rbot1_data = rbot1.data<float>();
    float * rbot2_data = rbot2.data<float>();

    int pwidthheight = paddedbottomwidth * paddedbottomheight;

    blob_rearrange_ongpu(input1_data,rbot1_data,batchSize,nInputPlane,nInputCols,nInputRows,inputWidthHeight,pad_size,pwidthheight);

    blob_rearrange_ongpu(input2_data,rbot2_data,batchSize,nInputPlane,nInputCols,nInputRows,inputWidthHeight,pad_size,pwidthheight);

    // CorrelationLayerBackward

    CorrelateDataBackward_ongpu(rbot1_data,rbot2_data,gradOutput_data,gradInput1_data,gradInput2_data,batchSize,nOutputCols,nOutputRows,nOutputPlane,max_displacement,neighborhood_grid_radius_,neighborhood_grid_width_,kernel_radius_,stride1,stride2,nInputCols,nInputRows,paddedbottomwidth,paddedbottomheight,nInputPlane,pad_size,corr_type_multiply);
}

File: correlation_package/correlation_package/corr_cuda.py
import torch
import math
import numpy as np
import torch.nn as nn
from torch.autograd import Function
from correlation_package_lib import corr_cuda

class correlationFunction(Function):
    @staticmethod
    def forward(ctx, pad_size, kernel_size, max_displacement, stride1, stride2, corr_multiply, input1, input2):
        ctx.pad_size=pad_size
        ctx.kernel_size=kernel_size
        ctx.max_displacement=max_displacement
        ctx.stride1=stride1
        ctx.stride2=stride2
        ctx.corr_multiply=corr_multiply
        ctx.input1=input1
        ctx.input2=input2

        B,C,H,W = input1.size()

        paddedbottomheight = H + 2 * pad_size
        paddedbottomwidth = W + 2 * pad_size

        kernel_radius_ = (kernel_size - 1) / 2
        border_size_ = max_displacement + kernel_radius_
        nOutputCols = int((paddedbottomwidth - border_size_ * 2.) / stride1+0.5)
        nOutputRows = int((paddedbottomheight - border_size_ * 2.) / stride1+0.5)
        neighborhood_grid_radius_ = max_displacement // stride2
        neighborhood_grid_width_ = neighborhood_grid_radius_ * 2 + 1
        nOutputPlane = neighborhood_grid_width_ * neighborhood_grid_width_

        rbot1 = torch.zeros([B, C, paddedbottomheight, paddedbottomwidth], dtype=torch.float32, device=input1.device)
        rbot2 = torch.zeros([B, C, paddedbottomheight, paddedbottomwidth], dtype=torch.float32, device=input1.device)
        output = torch.zeros([B, nOutputPlane, nOutputRows, nOutputCols], dtype=torch.float32, device=input1.device)

        corr_cuda().forward(input1, input2,
                        rbot1, rbot2,
                        output,
                        pad_size,
                        kernel_size,
                        max_displacement,
                        stride1,
                        stride2,
                        corr_multiply,
                        B,
                        C,
                        H,
                        W)

        return output

    @staticmethod
    def backward(ctx, grad_output):
        pad_size=ctx.pad_size
        kernel_size=ctx.kernel_size
        max_displacement=ctx.max_displacement
        stride1=ctx.stride1
        stride2=ctx.stride2
        corr_multiply=ctx.corr_multiply
        input1=ctx.input1
        input2=ctx.input2

        B,C,H,W = input1.size()

        paddedbottomheight = H + 2 * pad_size;
        paddedbottomwidth = W + 2 * pad_size;

        rbot1 = torch.zeros([B, C, paddedbottomheight, paddedbottomwidth], dtype=torch.float32, device=input1.device)
        rbot2 = torch.zeros([B, C, paddedbottomheight, paddedbottomwidth], dtype=torch.float32, device=input1.device)
        
        grad_input1 = torch.zeros_like(input1)
        grad_input2 = torch.zeros_like(input2)

        corr_cuda().backward(input1, input2,
                        rbot1, rbot2,
                        grad_output,
                        grad_input1,
                        grad_input2,
                        pad_size,
                        kernel_size,
                        max_displacement,
                        stride1,
                        stride2,
                        corr_multiply,
                        B,C,H,W)

        return None, None, None, None, None, None, grad_input1, grad_input2

class Correlation(nn.Module):

    def __init__(self, pad_size=3, kernel_size=3, max_displacement=20,
                 stride1=1, stride2=1, corr_multiply=1):
        super(Correlation, self).__init__()
        self.pad_size = pad_size
        self.kernel_size = kernel_size
        self.max_displacement = max_displacement
        self.stride1 = stride1
        self.stride2 = stride2
        self.corr_multiply = corr_multiply

    def forward(self, input1, input2):
        return correlationFunction.apply(self.pad_size, self.kernel_size, self.max_displacement, self.stride1, self.stride2, self.corr_multiply, input1, input2)






